python/tvm/contrib/peak.py

```diff
@@ -0,0 +1,332 @@
+
+        a = ib.allocate(dtype, (1), name='a', scope='local')
+        b = ib.allocate(dtype, (1), name='b', scope='local')
+
+        a[0] = outs[0].vload(idx, dtype)
+        b[0] = outs[0].vload(idx, dtype)
+
+        if base_type.find('float') != -1:
+            mad_func = (lambda x, y:
+                        tvm.call_pure_extern(dtype, 'mad', x, x, y))
+        else:
+            c = ib.allocate(dtype, (1), name='c', scope='local')
+            c[0] = outs[0].vload(idx, dtype)
+            mad_func = (lambda x, y: y * y + x)
+
+        for i in range(item_per_thread // 4 // lanes):
+            a[0] = mad_func(a[0], b[0])
+            b[0] = mad_func(b[0], a[0])
+
+        ib.emit(outs[0].vstore(idx, b[0]))
+        return ib.get()
+
+    y = tvm.extern((n,), [], extern, name="y", dtype=dtype)
+    s = tvm.create_schedule(y.op)
+
+    try:
+        func = tvm.build(s, [y], target, target_host=target_host)
+        func = _convert_to_remote(func, remote)
+        time_f = func.time_evaluator(func.entry_name, ctx, number=n_times)
+        y = tvm.nd.empty((n,), dtype=dtype, ctx=ctx)
+        time = time_f(y).mean
+    except tvm._ffi.base.TVMError as e:
+        # build error (occur when device does not support half)
+        return 0
+
+    return 1.0 * (total_item * item_per_thread) / (1 << 30) / time
+
+def test_compute_all_types(total_item, item_per_thread, n_times, target, target_host, remote, ctx, verbose=True):
+    """ test memory bandwidth for all types
+
+    Parameters
+    ----------
+    total_item: int
+        number of elements in input array
+    item_per_thread: int
+        number of elements each thread accmulates
+    n_times: int
+        number of runs for averaging
+    target: :any:`tvm.target.Target`
+        the target and option of the compilation.
+    target_host : str or :any:`tvm.target.Target`
+        host compilation target
+    remote: tvm.contrib.rpc.RPCSession
+        remote rpc session
+    ctx: TVMcontext
+        the context of array
+    verbose: bool
+        whether outputs immediate result
+
+    Returns
+    -------
+    result: list
+        a list of (type_name, GFLOPS/GIOPS) pairs
+    """
+    result = []
+    for base_type in ["int", "float"]:
+        for bits in [16, 32, 64]:
+            for lanes in [1, 2, 4, 8, 16]:
+                if base_type == 'int' and bits != 32:  # only test int32
+                    continue
+
+                speed = test_compute_mad(total_item, item_per_thread,
+                                         base_type, bits, lanes, target,
+                                         target_host, remote, ctx, n_times)
+                type_name = get_name(base_type, bits)
+                result.append(["%s%d" % (type_name, lanes), speed])
+
+                unit = "GFLOPS" if base_type == "float" else "GIOPS"
+
+                if verbose:
+                    logging.info("\t%-10s %.2f %s" % (result[-1][0], result[-1][1], unit))
```

Comment:
when you use logging module, always use printf style
```logging.info(format, x, y, z) ``` instead of  ```logging.info(format %(x, y, z))```
x, y, z will be lazily evaluated in the former case